### PR TITLE
Support InsertStmt in new plan

### PR DESCRIPTION
**Pull request by shenli**
@coocood @qiuyesuifeng

### CODE REVIEW

File: optimizer/typeinferer.go [hunk 1]
```diff
@@ -181,8 +181,10 @@ func (v *typeInferrer) binaryOperation(x *ast.BinaryOperationExpr) {
 			x.Type = types.NewFieldType(xTp)
 		}
 	}
-	x.Type.Charset = charset.CharsetBin
-	x.Type.Collate = charset.CollationBin
+	if x.Type != nil {
```
Comment (coocood) [inline optimizer/typeinferer.go]: If you infer `Values` expression, `x.Type` should never be nil.

File: executor/executor_write.go [hunk 2]
```diff
@@ -342,3 +358,365 @@ func (e *DeleteExec) Fields() []*ast.ResultField {
...
+	Setlist     []*ast.Assignment
+	OnDuplicate []*ast.Assignment
+	fields      []*ast.ResultField
+
+	IsReplace bool
+	Priority  int
+}
+
+// Next implements Executor Next interface.
+func (e *InsertExec) Next() (*Row, error) {
```
Comment (coocood) [inline executor/executor_write.go]: Add `done` field to make sure it is executed only once.

File: executor/executor_write.go [hunk 3]
```diff
@@ -342,3 +358,365 @@ func (e *DeleteExec) Fields() []*ast.ResultField {
...
+				} else {
+					vals[j] = defaultVals[cols[j].Name.L]
+				}
+			} else {
+				vals[j], err = evaluator.Eval(e.ctx, expr)
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+			}
+		}
```
Comment (coocood) [inline executor/executor_write.go]: This block is too deep, extract to a function.

File: executor/builder.go [hunk 4]
```diff
@@ -333,3 +336,39 @@ func (b *executorBuilder) buildDelete(v *plan.Delete) Executor {
...
+		IsReplace:   v.IsReplace,
+		Priority:    v.Priority,
+	}
+	if v.SelectPlan != nil {
+		insert.SelectExec = b.build(v.SelectPlan)
+	}
+	// Get Table
+	ts, ok := v.Table.TableRefs.Left.(*ast.TableSource)
+	if !ok {
+		b.err = errors.Errorf("Can not get table")
```
Comment (qiuyesuifeng) [inline executor/builder.go]: errors.New

**Comment by shenli** [comment]
PTAL @coocood @qiuyesuifeng

**Comment by coocood** [comment]
LGTM

File: executor/executor_write.go [hunk 5]
```diff
@@ -342,3 +358,376 @@ func (e *DeleteExec) Fields() []*ast.ResultField {
...
+			return nil, errors.Trace(err)
+		}
+	}
+	return
+}
+
+func (e *InsertExec) getRow(cols []*column.Col, list []ast.ExprNode, defaultVals map[string]interface{}) ([]interface{}, error) {
+	vals := make([]interface{}, len(list))
+	var err error
+	for j, expr := range list {
```
Comment (qiuyesuifeng) [inline executor/executor_write.go]: How about using i instead of j?

**Comment by shenli** [comment]
@qiuyesuifeng PTAL

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**